From 49d59cdd58b8332f425ff05f3f26de1c7e6ec0dd Mon Sep 17 00:00:00 2001
From: Joshua Haberman <haberman@google.com>
Date: Fri, 21 Jun 2024 11:12:25 -0700
Subject: [PATCH 1/7] Make our Python source wheel use the version script to
 restruct symbol visibility.

We generally want our extension module to hide all symbols except a single entry point, `PyInit__message()`.  We accomplish this through a `version_script.lds`, which instructs the linker to hide all symbols except our single entry point.

Currently, only our Bazel-based build is using `version_script.lds`.  This CL makes our Python source package also use `version_script.lds`, so that it can produce binary wheel that are more similar to the Bazel-built ones that we release.

PiperOrigin-RevId: 645445353
---
 python/BUILD.bazel      | 2 ++
 python/dist/BUILD.bazel | 1 +
 python/dist/MANIFEST.in | 3 ++-
 python/dist/setup.py    | 2 +-
 4 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/python/BUILD.bazel b/python/BUILD.bazel
index 325a56b..0023876 100644
--- a/python/BUILD.bazel
+++ b/python/BUILD.bazel
@@ -181,6 +181,8 @@ filegroup(
     # end:google_only
 )
 
+exports_files(["version_script.lds"])
+
 py_extension(
     name = "_message",
     srcs = [":message_srcs"],
diff --git a/python/dist/BUILD.bazel b/python/dist/BUILD.bazel
index 03c44c0..861c558 100644
--- a/python/dist/BUILD.bazel
+++ b/python/dist/BUILD.bazel
@@ -237,6 +237,7 @@ pkg_files(
     srcs = [
         "//:LICENSE",
         "//python:message_srcs",
+        "//python:version_script.lds",
         "//upb:source_files",
         "//upb/base:source_files",
         "//upb/hash:source_files",
diff --git a/python/dist/MANIFEST.in b/python/dist/MANIFEST.in
index 1b61936..f243c10 100644
--- a/python/dist/MANIFEST.in
+++ b/python/dist/MANIFEST.in
@@ -1,2 +1,3 @@
 global-include *.h
-global-include *.inc
\ No newline at end of file
+global-include *.inc
+include python/version_script.lds
diff --git a/python/dist/setup.py b/python/dist/setup.py
index 534e3fb..2938a1b 100755
--- a/python/dist/setup.py
+++ b/python/dist/setup.py
@@ -33,7 +33,7 @@ def GetVersion():
 
 
 current_dir = os.path.dirname(os.path.abspath(__file__))
-extra_link_args = []
+extra_link_args = ["-Wl,--version-script=python/version_script.lds"]
 
 if sys.platform.startswith('win'):
   extra_link_args = ['-static']
